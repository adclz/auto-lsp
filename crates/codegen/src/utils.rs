pub(crate) fn sanitize_string(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result.replace(" ", "_")
    }
}

pub(crate) fn sanitize_string_to_pascal(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result
            .replace(" ", "_")
            .split('_')
            .map(|word| {
                let mut chars = word.chars();
                match chars.next() {
                    Some(first) => format!(
                        "{}{}",
                        first.to_ascii_uppercase(),
                        chars.collect::<String>()
                    ),
                    None => String::new(),
                }
            })
            .collect()
    }
}

pub static PUNCTUATION: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "{" => "LeftCurly",
    "}" => "RightCurly",
    "(" => "LeftParen",
    ")" => "RightParen",
    "[" => "LeftBracket",
    "]" => "RightBracket",
    "," => "Comma",
    ":" => "Colon",
    ";" => "Semicolon",
    "." => "Dot",
    "@" => "At",
    "!" => "Bang",
    "#" => "Hash",
    "$" => "Dollar",
    "%" => "Percent",
    "^" => "Caret",
    "&" => "Ampersand",
    "*" => "Star",
    "-" => "Minus",
    "_" => "Underscore",
    "+" => "Plus",
    "=" => "Equal",
    ">" => "Greater",
    "<" => "Less",
    "|" => "Pipe",
    "~" => "Tilde",
    "/" => "Slash",
    "\\" => "Backslash",
    "//" => "SlashSlash",
    "//=>" => "SlashSlashEqual",
    "//=" => "SlashSlashEqual",
    "/=" => "SlashEqual",
    "/>" => "SlashGreater",
    "/?" => "SlashQuestion",
    "/??" => "SlashNullish",
    "/*" => "SlashStar",
    "*/" => "StarSlash",
    "+++" => "PlusPlusPlus",
    "!!" => "BangBang",
    "!!=" => "BangBangEqual",
    "!!?" => "BangBangQuestion",
    "!!??" => "BangBangNullish",
    "!!???" => "BangBangNullishQuestion",
    "?" => "Question",
    "->" => "Arrow",
    "=>" => "FatArrow",
    "++" => "PlusPlus",
    "--" => "MinusMinus",
    "&&" => "And",
    "||" => "Or",
    "==" => "EqualEqual",
    "!=" => "NotEqual",
    ">=" => "GreaterEqual",
    "<=" => "LessEqual",
    "===" => "TripleEqual",
    "!==" => "NotTripleEqual",
    "<<" => "ShiftLeft",
    ">>" => "ShiftRight",
    ">>>" => "ShiftRightUnsigned",
    "+=" => "PlusEqual",
    "-=" => "MinusEqual",
    "*=" => "StarEqual",
    "%=" => "PercentEqual",
    "&=" => "AmpersandEqual",
    "|=" => "PipeEqual",
    "^=" => "CaretEqual",
    "&&=" => "AndEqual",
    "||=" => "OrEqual",
    "??=" => "NullishEqual",
    "??" => "Nullish",
    "???" => "NullishQuestion",
    "**" => "StarStar",
    "**=" => "StarStarEqual",
    "<>" => "LessGreater",
    "<=>" => "LessGreaterEqual",
};

pub static RUST_KEYWORDS: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "abstract" => "Abstract",
    "as" => "As",
    "async" => "Async",
    "await" => "Await",
    "break" => "Break",
    "const" => "Const",
    "continue" => "Continue",
    "crate" => "Crate",
    "dyn" => "Dyn",
    "else" => "Else",
    "enum" => "Enum",
    "extern" => "Extern",
    "false" => "False",
    "final" => "Final",
    "fn" => "Fn",
    "for" => "For",
    "if" => "If",
    "impl" => "Impl",
    "in" => "In",
    "let" => "Let",
    "loop" => "Loop",
    "match" => "Match",
    "mod" => "Mod",
    "move" => "Move",
    "mut" => "Mut",
    "pub" => "Pub",
    "ref" => "Ref",
    "return" => "Return",
    "self" => "Self",
    "static" => "Static",
    "struct" => "Struct",
    "super" => "Super",
    "trait" => "Trait",
    "true" => "True",
    "type" => "Type",
    "unsafe" => "Unsafe",
    "use" => "Use",
    "where" => "Where",
    "while" => "While",
    "with" => "With",
    "yield" => "Yield",
    "None" => "_None",
    "Some" => "_Some",
    "Ok" => "_Ok",
    "Err" => "_Err",
    "Result" => "_Result",
    "Option" => "_Option",
    "Vec" => "_Vec",
};
