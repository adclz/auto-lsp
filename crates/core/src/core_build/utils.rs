/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

pub(crate) fn intersecting_ranges(
    range1: &std::ops::Range<usize>,
    range2: &tree_sitter::Range,
) -> bool {
    range1.start <= range2.start_byte && range1.end >= range2.end_byte
}

pub(crate) fn tree_sitter_range_to_lsp_range(range: &tree_sitter::Range) -> lsp_types::Range {
    let start = range.start_point;
    let end = range.end_point;
    lsp_types::Range {
        start: lsp_types::Position {
            line: start.row as u32,
            character: start.column as u32,
        },
        end: lsp_types::Position {
            line: end.row as u32,
            character: end.column as u32,
        },
    }
}
