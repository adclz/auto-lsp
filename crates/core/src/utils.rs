#[macro_export]
macro_rules! dispatch_once {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                return n.$method($($param),*);
            };
        )*
    };
}

#[macro_export]
macro_rules! dispatch {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                n.$method($($param),*)?;
            };
        )*
    };
}