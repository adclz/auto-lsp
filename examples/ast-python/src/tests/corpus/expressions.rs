/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

use super::utils::Result;
use crate::snap;

#[test]
fn identifiers_with_greek_letters() -> Result {
    snap!(
        r#"
ψ1 = β_γ + Ψ_5
"#
    )
}

#[test]
fn subscript_expressions() -> Result {
    snap!(
        r#"
a[1]
b[2, 3]
c[4, 5,]
"#
    )
}

#[test]
fn subscript_slice_expressions() -> Result {
    snap!(
        r#"
a[:]
b[5:]
b[5:6, ...]
c[::]
"#
    )
}

#[test]
fn attribute_references() -> Result {
    snap!(
        r#"
a.b.c
"#
    )
}

#[test]
fn await_expressions() -> Result {
    snap!(
        r#"
await i(j, 5)
return await i(j, 5)
async def region_exists(region: str) -> bool:
    return region in await all_regions()

assert await a(b) == c
"#
    )
}

#[test]
fn call_expressions() -> Result {
    snap!(
        r#"
__a__()
b(1)
c(e, f=g)
i(j, 5,)
"#
    )
}

#[test]
fn print_used_as_an_identifier() -> Result {
    snap!(
        r#"
print()
print(a)
print(a, b=c)
print(d, e)
print(d, *e)
print(*f, **g,)
a(print)
"#
    )
}

#[test]
fn print_used_as_a_parameter() -> Result {
    snap!(
        r#"
def a(print):
  b
def a(printer=print):
  c
def a(*print):
  b
def a(**print):
  b
def print():
  a
"#
    )
}

#[test]
fn exec_used_as_an_identifier() -> Result {
    snap!(
        r#"
exec("print \"'%s' has %i characters\" % (public_function(), len(public_function()))", {"__builtins__" : None}, safe_dict)
exec("""exec _code_ in _globs_, _locs_""")
"#
    )
}

#[test]
fn async_await_used_as_identifiers() -> Result {
    snap!(
        r#"
async = 4
await = 5
print async, await
"#
    )
}

#[test]
fn calls_with_splats() -> Result {
    snap!(
        r#"
a(*())
a(**{})
a(*b)
c(d, *e, **g)
"#
    )
}

#[test]
fn math_operators() -> Result {
    snap!(
        r#"
a + b * c ** d - e / 5
a // 2
-5
+x
~x
"#
    )
}

#[test]
fn binary_addition_subtraction_with_floats() -> Result {
    snap!(
        r#"
.1-.0
.1+.0
.1-0
.1+0

1-.0
1+.0
"#
    )
}

#[test]
fn power_operator_precedence() -> Result {
    snap!(
        r#"
2**2**3
-2**2
"#
    )
}

#[test]
fn operator_precedence() -> Result {
    snap!(
        r#"
a() + b[c] * c.d.e
"#
    )
}

#[test]
fn bitwise_operators() -> Result {
    snap!(
        r#"
a << b | c >> d & e ^ f
"#
    )
}

#[test]
fn boolean_operators() -> Result {
    snap!(
        r#"
a or b and c
not d
not a and b or c
a and not b and c
"#
    )
}

#[test]
fn comparison_operators() -> Result {
    snap!(
        r#"
a < b <= c == d >= e > f
not a == b or c == d
a not in b
a is not b
a is b and c != d
a <> b
"#
    )
}

#[test]
fn assignments() -> Result {
    snap!(
        r#"
a = 1
a, b = 1, 2
a, *c = 1, 2, 3
a, = 1, 2
a[b] = c = d
a, *b.c = d
"#
    )
}

#[test]
fn assignments_with_type_annotations() -> Result {
    snap!(
        r#"
tail_leaves: List[Leaf] = []
"#
    )
}

#[test]
fn augmented_assignments() -> Result {
    snap!(
        r#"
a += 1
b >>= 2
c //= 1
d @= e
f -= 2
g %= 2
h /= 5
i *= 2.2
j &= 1
k ^= 0
l <<= 2
m |= k
n **= 3
"#
    )
}

#[test]
fn named_expressions() -> Result {
    snap!(
        r#"
a := x
(y := f(x))
foo(x=(y := f(x)))
y0 = (y1 := f(x))
def foo(answer=(p := 42)):
  return answer;
def foo(answer: (p := 42) = 5):
  return answer;
foo(x := 3, cat='vector')
(z := (y := (x := 0)))
"#
    )
}

#[test]
fn yield_expressions() -> Result {
    snap!(
        r#"
def example():
  yield
  yield 1
  x = yield 2
  yield from a
  yield from (yield from (x for x in range(1, 10)))
"#
    )
}

#[test]
fn lambdas() -> Result {
    snap!(
        r#"
lambda b, c: d("e" % f)
lambda: True
lambda a, b = c, *d, **e: a
lambda (a, b): (a, b)
"#
    )
}

#[test]
fn tuples_with_splats() -> Result {
    snap!(
        r#"
(foo, *bar, *baz)
"#
    )
}
