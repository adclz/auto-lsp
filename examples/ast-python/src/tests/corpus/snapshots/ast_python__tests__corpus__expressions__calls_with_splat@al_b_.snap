---
source: src/tests/corpus/expressions.rs
expression: module
---
Module {
    children: [
        SimpleStatement(
            ExpressionStatement(
                ExpressionStatement {
                    children: [
                        Expression(
                            PrimaryExpression(
                                Call(
                                    Call {
                                        function: Identifier(
                                            Identifier {
                                                [RANGE]
                                            },
                                        ),
                                        arguments: ArgumentList(
                                            ArgumentList {
                                                children: [
                                                    ListSplat(
                                                        ListSplat {
                                                            children: Identifier(
                                                                Identifier {
                                                                    [RANGE]
                                                                },
                                                            ),
                                                            [RANGE]
                                                        },
                                                    ),
                                                ],
                                                [RANGE]
                                            },
                                        ),
                                        [RANGE]
                                    },
                                ),
                            ),
                        ),
                    ],
                    [RANGE]
                },
            ),
        ),
    ],
    [RANGE]
}
